src/openrct2/object/TerrainSurfaceObject.cpp

```diff
@@ -136,24 +138,44 @@ void TerrainSurfaceObject::ReadJson(IReadObjectContext* context, json_t& root)
     PopulateTablesFromJson(context, root);
 }
 
-uint32_t TerrainSurfaceObject::GetImageId(
-    const CoordsXY& position, int32_t length, int32_t rotation, int32_t offset, bool grid, bool underground) const
+ImageId TerrainSurfaceObject::GetImageId(
+    const CoordsXY& position, uint8_t length, uint8_t rotation, uint8_t offset, bool grid, bool underground) const
 {
-    uint32_t result = (underground ? DefaultUndergroundEntry : (grid ? DefaultGridEntry : DefaultEntry));
+    uint32_t result = DefaultEntry;
+    const std::vector<SpecialEntry>* entries = &SpecialEntries;
+    if (underground)
+    {
+        result = DefaultUndergroundEntry;
+        entries = &SpecialEntriesUnderground;
+    }
+    else if (grid)
+    {
+        result = DefaultGridEntry;
+        entries = &SpecialEntriesGrid;
+    }
+
+    int32_t x = position.x >> 5;
+    int32_t y = position.y >> 5;
```

Comment:
Please change this to use `TileCoordsXY(position)`, rather than manually dividing by 32.